spring-integration-core/src/main/java/org/springframework/integration/endpoint/MessageProducerSupport.java

```diff
@@ -187,18 +188,42 @@ protected void sendMessage(Message<?> message) {
 			this.messagingTemplate.send(getOutputChannel(), message);
 		}
 		catch (RuntimeException e) {
-			MessageChannel errorChannel = getErrorChannel();
-			if (errorChannel != null) {
-				this.messagingTemplate.send(errorChannel, this.errorMessageStrategy.buildErrorMessage(e,
-						getErrorMessageAttributes(message)));
-			}
-			else {
+			if (!sendErrorMessageIfNecessary(message, e)) {
 				throw e;
 			}
 		}
 	}
 
 	/**
+	 * Send an error message based on the exception and message.
+	 * @param message the message.
+	 * @param exception the exception.
+	 * @return true if the error channel is available and message sent.
+	 * @since 4.3.10
+	 */
+	protected boolean sendErrorMessageIfNecessary(Message<?> message, RuntimeException exception) {
+		MessageChannel errorChannel = getErrorChannel();
+		if (errorChannel != null) {
+			this.messagingTemplate.send(errorChannel, buildErrorMessage(message, exception));
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Build an error message for the exception and message using the configured
+	 * {@link ErrorMessageStrategy}.
```

Comment:
Or let's make this method `final` or mentioning of the `ErrorMessageStrategy` does not make sense.
The implementor can simply override this method on his own.
And `ErrorMessageStrategy` premise may be eliminated.